### PR TITLE
Tg blas3

### PR DESCRIPTION
**Pull request by ZuseZ4**
TODO: mabye set writeonly

### CODE REVIEW

**Comment by ZuseZ4** [comment]
@vchuravy In a first step I would start calling dot from julia side. The generated code for handle_dot is ~500 Loc.
I can set most of the generated variables to const, only emit handle_dot code for now, and eventually write a few more docs. Any other thoughts?
Especially, can you please check the Attributes I add in the generated code?
For now I add ArgMemOnly, ReadOnly and NoCapture in places where it should be suitable.

**Comment by ZuseZ4** [comment]
Jullia is not quite happy yet:
```
define internal { float } @diffejulia_dot_9525_inner.1(i64 signext %0, {} addrspace(10)* nonnull align 16 dereferenceable(40) %1, i64 signext %2, {} addrspace(10)* nonnull align 16 dereferenceable(40) %3, i64 signext %4) local_unnamed_addr #6 !dbg !47 {
entry:
  %toreturn = alloca float, align 4
  %5 = alloca i64, align 16
  %6 = bitcast i64* %5 to i8*
  %7 = alloca i64, align 16
  %8 = bitcast i64* %7 to i8*
  %9 = alloca i64, align 16
  %10 = bitcast i64* %9 to i8*
  %11 = call {}*** @julia.get_pgcstack() #7
  call void @llvm.lifetime.start.p0i8(i64 noundef 8, i8* noundef nonnull %6) #7
  store i64 %0, i64* %5, align 16, !dbg !48, !tbaa !18, !noalias !53
  call void @llvm.lifetime.start.p0i8(i64 noundef 8, i8* noundef nonnull %8) #7
  store i64 %2, i64* %7, align 16, !dbg !48, !tbaa !18, !noalias !53
  call void @llvm.lifetime.start.p0i8(i64 noundef 8, i8* noundef nonnull %10) #7
  store i64 %4, i64* %9, align 16, !dbg !48, !tbaa !18, !noalias !53
  %12 = addrspacecast {} addrspace(10)* %1 to {} addrspace(11)*, !dbg !56
  %13 = call nonnull {}* @julia.pointer_from_objref({} addrspace(11)* %12) #8, !dbg !56
  %14 = bitcast {}* %13 to i8**, !dbg !56
  %15 = load i8*, i8** %14, align 8, !dbg !56, !tbaa !30, !nonnull !4
  %"'il_phi" = phi i8* , !dbg !56
  %16 = ptrtoint i8* %15 to i64, !dbg !56
  %17 = addrspacecast {} addrspace(10)* %3 to {} addrspace(11)*, !dbg !56
  %18 = call nonnull {}* @julia.pointer_from_objref({} addrspace(11)* %17) #8, !dbg !56
  %19 = bitcast {}* %18 to i8**, !dbg !56
  %20 = load i8*, i8** %19, align 8, !dbg !56, !tbaa !30, !nonnull !4
  %"'il_phi1" = phi i8* , !dbg !56
  %21 = ptrtoint i8* %20 to i64, !dbg !56
  %22 = call float @sdot_64_(i8* noundef nonnull %6, i64 %16, i8* noundef nonnull %8, i64 %21, i8* noundef nonnull %10) #7 [ "jl_roots"({} addrspace(10)* null, {} addrspace(10)* %3, {} addrspace(10)* null, {} addrspace(10)* %1, {} addrspace(10)* null) ], !dbg !57
  store float %22, float* %toreturn, align 4, !dbg !58
  br label %invertentry, !dbg !58

allocsForInversion:                               ; No predecessors!
  %23 = alloca float, align 4

invertentry:                                      ; preds = %entry
}

  %22 = call float @sdot_64_(i8* noundef nonnull %6, i64 %16, i8* noundef nonnull %8, i64 %21, i8* noundef nonnull %10) #7 [ "jl_roots"({} addrspace(10)* null, {} addrspace(10)* %3, {} addrspace(10)* null, {} addrspace(10)* %1, {} addrspace(10)* null) ], !dbg !33
julia: /u/drehwald/prog/enzyme_blas/enzyme/Enzyme/DiffeGradientUtils.cpp:203: llvm::Value* DiffeGradientUtils::diffe(llvm::Value*, llvm::IRBuilder<>&): Assertion `0 && "getting diffe of constant value"' failed.

```

File: enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp [hunk 1]
```diff
@@ -717,11 +727,1059 @@ static void emitDerivatives(const RecordKeeper &recordKeeper, raw_ostream &os) {
...
+     << "      Mode == DerivativeMode::ReverseModeGradient) {\n"
+     << "    Value *alloc = nullptr;\n"
+     << "    if (byRef) {\n"
+     << "      alloc = allocationBuilder.CreateAlloca(fpType);\n"
+     << "    }\n\n";
+  if (hasDiffeRetVal) {
+    os << "    Value *dif = diffe(&call, Builder2);\n";
+  }
+
+  // TODO: adj_ args
```
Comment (wsmoses) [inline enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp:1588]: Cleanup some of these comments?

Comment (ZuseZ4) [reply enzyme/tools/enzyme-tblgen/enzyme-tblgen.cpp]: done

File: enzyme/BCLoad/BCLoader.cpp [hunk 2]
```diff
@@ -27,11 +27,20 @@ bool provideDefinitions(Module &M) {
...
-              F.getName().substr(0, F.getName().size() - strlen(postfix)).str();
+        if (str == "sdot" || str == "ddot") {
+          llvm::errs() << "SKIPPING DOT \n\n\n\n";
+          continue; // don't handle tablegen'd ones
+        }
+      } else if (F.getName().endswith(postfix)) {
+        auto blasName =
+            F.getName().substr(0, F.getName().size() - strlen(postfix)).str();
+        if (blasName == "sdot" || blasName == "ddot") {
+          llvm::errs() << "SKIPPING DOT \n\n\n\n";
```
Comment (wsmoses) [inline enzyme/BCLoad/BCLoader.cpp:40]: Are we also able to remove the complex dot as well?

Comment (ZuseZ4) [reply enzyme/BCLoad/BCLoader.cpp:40]: sure! will also remove the print once I got the julia testcases passing.

**Comment by ZuseZ4** [comment]
@wsmoses Should be good to go :)

File: enzyme/BCLoad/BCLoader.cpp [hunk 3]
```diff
@@ -27,11 +27,24 @@ bool provideDefinitions(Module &M) {
...
       std::string str;
-      if (strlen(postfix) == 0)
+      if (strlen(postfix) == 0) {
         str = F.getName().str();
-      else if (F.getName().endswith(postfix)) {
-        str = "cblas_" +
-              F.getName().substr(0, F.getName().size() - strlen(postfix)).str();
+        for (auto ignoreFnc : ignoreFunctions) {
+          if (str == ignoreFnc) {
+            llvm::errs() << "SKIPPING: " << str << "\n\n\n\n";
```
Comment (wsmoses) [inline enzyme/BCLoad/BCLoader.cpp:34]: remove the prints

File: enzyme/Enzyme/AdjointGenerator.h [hunk 4]
```diff
@@ -11318,12 +10594,22 @@ class AdjointGenerator
       return;
     }
 
+    // TODO: BLAS re-adding
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:10597]: cleanup?

File: enzyme/BCLoad/BCLoader.cpp [hunk 5]
```diff
@@ -27,11 +27,22 @@ bool provideDefinitions(Module &M) {
     int index = 0;
     for (auto postfix : {"", "_", "_64_"}) {
       std::string str;
-      if (strlen(postfix) == 0)
+      if (strlen(postfix) == 0) {
         str = F.getName().str();
-      else if (F.getName().endswith(postfix)) {
-        str = "cblas_" +
-              F.getName().substr(0, F.getName().size() - strlen(postfix)).str();
+        for (auto ignoreFnc : ignoreFunctions) {
```
Comment (wsmoses) [inline enzyme/BCLoad/BCLoader.cpp:32]: Change this to if ignoreFunctions.count(funcName.str())

**Pull request merged**